README.md
Retrieve data from Google Analytics
=================
[![Latest Stable Version](https://poser.pugx.org/spatie/analytics-reports/version.png)](https://packagist.org/packages/spatie/analytics-reports)
[![Total Downloads](https://poser.pugx.org/spatie/analytics-reports/downloads.svg)](https://packagist.org/packages/spatie/analytics-reports)
[![License](https://poser.pugx.org/spatie/analytics-reports/license.png)](https://packagist.org/packages/spatie/analytics-reports)

This is an opinionated Laravel 4 package to retrieve Google Analytics data. 

If you're using Laravel 5, [take a look at this package](https://github.com/spatie/laravel-analytics).

Spatie is a webdesign agency in Antwerp, Belgium. You'll find an overview of all our open source projects [on our website](https://spatie.be/opensource).


## Installation

This package can be installed through Composer. 

```
composer require spatie/analytics-reports
```

You must install this service provider.

```php

// app/config/app.php

'providers' => [
    '...',
    'Spatie\AnalyticsReports\AnalyticsReportsServiceProvider'
];
```

This package also comes with a facade, which provides an easy way to call the the class.


```php

// app/config/app.php

'aliases' => array(
	...
	'AnalyticsReports' => 'Spatie\AnalyticsReports\AnalyticsReportsFacade',
)
```

Although the composer.json of this package specifies that "google/apiclient" AND "thujohn/analytics" must be pulled in as well, it sometimes fails to do so. If you encouter this problem as well include these lines in your composer.json as well:
```js
{
    "require": {
    ...
        "google/apiclient" : "1.1.*",
        "thujohn/analytics": "dev-master",
	...
	}
}
```

You can publish the config file of the package using artisan

```bash
php artisan config:publish spatie/analytics-reports
```
After the config file has been published you'll manually have to move it to your app's config-folder. (Hopefully this step won't be necessary in a next version)

To move the config file copy the ```config.php``` file from 
```
/app/config/packages/spatie/analytics-reports/
```

and paste it to your ```/app/config``` folder and rename it to ```analyticsReports.php```

In the config file you can specify two values:
- siteId: the Google site id, something in the form of ga:xxxxxxxx
- cacheLifeTime: the amount of minutes the Google API responses will be cached. If you set this value to zero, the responses won't be cached at all.

Internally this package uses [thujohn/analytics-l4](https://github.com/thujohn/analytics-l4) to authenticate with Google. So in order to use this package you must also follow [their installation instructions](https://github.com/thujohn/analytics-l4#installation).



## Usage


When the installation is done you can easily retrieve Analytics data. Mostly all methods will return an `Illuminate\Support\Collection`-instance.


Here is an example to retrieve visitors and pageview data for the last seven days.
```php
/*
* $analyticsData now contains a Collection with 3 columns: "date", "visitors" and "pageViews"
*/
$analyticsData = AnalyticsReports::getVisitorsAndPageViews(7)
```

Here's another example to get the 20 most visited pages of the last 365 days
```php
/*
* $analyticsData now contains a Collection with 2 columns: "url" and "pageViews"
*/
$analyticsData = AnalyticsReports::getMostVisitedPages(365, 20)
```
## Provided methods

###Visitors and Pageviews
These methods return a Collection with columns "date", "vistors" and "pageViews". When grouping by yearMonth, the first column will be called "yearMonth".
```php
    /**
     * Get the amount of visitors and pageviews
     *
     * @param int $numberOfDays
     * @param string $groupBy Possible values: date, yearMonth
     * @return Collection
     */
    public function getVisitorsAndPageViews($numberOfDays = 365, $groupBy = 'date')

    /**
     * Get the amount of visitors and pageviews for the given period
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param string $groupBy Possible values: date, yearMonth
     * @return Collection
     */
    public function getVisitorsAndPageViewsForPeriod($startDate, $endDate, $groupBy = 'date')
```    

###Keywords
These methods return a Collection with columns "keyword" and "sessions".
```php
   /**
     * Get the top keywords
     *
     * @param int $numberOfDays
     * @param int $maxResults
     * @return Collection
     */
    public function getTopKeywords($numberOfDays = 365, $maxResults = 30)

    /**
     * Get the top keywords for the given period
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param int $maxResults
     * @return Collection
     */
    public function getTopKeyWordsForPeriod($startDate, $endDate, $maxResults = 30)
```

###Referrers
These methods return a Collection with columns "url" and "pageViews".
```php
    /**
     * Get the top referrers
     *
     * @param int $numberOfDays
     * @param int $maxResults
     * @return Collection
     */
    public function getTopReferrers($numberOfDays = 365, $maxResults = 20)

    /**
     * Get the top referrers for the given period
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $maxResults
     * @return Collection
     */
    public function getTopReferrersForPeriod($startDate, $endDate, $maxResults)
``` 

###Browsers
These methods return a Collection with columns "browser" and "sessions".

If there are  more used browsers than the number specified in maxResults, then a new resultrow with browser-name "other" will be appended with a sum of all the remaining browsers.
```php
    /**
     * Get the top browsers
     *
     * @param int $numberOfDays
     * @param int $maxResults
     * @return Collection
     */
    public function getTopBrowsers($numberOfDays = 365, $maxResults = 6)
    
    /**
     * Get the top browsers for the given period
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $maxResults
     * @return Collection
     */
    public function getTopBrowsersForPeriod($startDate, $endDate, $maxResults) 
```     

###Most visited pages
These methods return a Collection with columns "url" and "pageViews".
```php
    /**
     * Get the most visited pages
     *
     * @param int $numberOfDays
     * @param int $maxResults
     * @return Collection
     */
    public function getMostVisitedPages($numberOfDays = 365, $maxResults = 20)
    
    /**
     * Get the most visited pages for the given period
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param int $maxResults
     * @return Collection
     */
    public function getMostVisitedPagesForPeriod($startDate, $endDate, $maxResults = 20)
```

###All other Google Analytics Queries
To perform all other GA queries use  ```performQuery```.  [Google's Core Reporting API](https://developers.google.com/analytics/devguides/reporting/core/v3/common-queries) provides more information on on which metrics and dimensions might be used. 
```php
    /**
     * Call the query method on the autenthicated client
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $metrics
     * @param array $others
     * @return mixed
     */
    public function performQuery($startDate, $endDate, $metrics, $others = array())
```    

##Convenience methods
```getSiteIdByUrl``` can be used to get the site id for the given url

```php
    /**
     * Returns the site id (ga:xxxxxxx) for the given url
     *
     * @param $url
     * @throws \Exception
     * @return string
     */
    public function getSiteIdByUrl($url)
```           

## About Spatie
Spatie is a webdesign agency in Antwerp, Belgium. You'll find an overview of all our open source projects [on our website](https://spatie.be/opensource).


CHANGELOG.md
# Changelog

All Notable changes to `spatie/analytics-reports` will be documented in this file

##1.0.2 2016-02-03
Bugfix: required semantic version of thujohn/analytics

##1.0.1 2015-06-04
Bugfix: remove double use of Cache namespace



src/config/config.php
<?php

return

    [
        /*
        |--------------------------------------------------------------------------
        | Analytics reports site id
        |--------------------------------------------------------------------------
        |
        | This site id is used to retrieve and display Google Analytics statistics
        | in the admin-section. Should be something like ga:xxxxxxxx.
        |
        */

        'siteId' => '',

        /*
        |--------------------------------------------------------------------------
        | Cache Lifetime
        |--------------------------------------------------------------------------
        |
        | The amount of minutes the Google API responses will be cached.
        | If you set this to zero, the responses won't be cached at all.
        |
        */

        'cacheLifetime' => 60 * 24 * 2,

    ];


src/config/analyticsReports.php
<?php return

    [
        /*
        |--------------------------------------------------------------------------
        | Analytics reports site id
        |--------------------------------------------------------------------------
        |
        | This site id is used to retrieve and display Google Analytics statistics
        | in the admin-section. Should be something like ga:xxxxxxxx.
        |
        */

        'siteId' => '',

        /*
        |--------------------------------------------------------------------------
        | Cache Lifetime
        |--------------------------------------------------------------------------
        |
        | The amount of minutes the Google API responses will be cached.
        | If you set this to zero, the responses won't be cached at all.
        |
        */

        'cacheLifetime' => 60 * 24 * 2

    ];

src/AnalyticsReportsServiceProvider.php
<?php

namespace Spatie\AnalyticsReports;

use Config;
use Illuminate\Support\ServiceProvider;

class AnalyticsReportsServiceProvider extends ServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = true;

    /**
     * Bootstrap the application events.
     *
     * @return void
     */
    public function boot()
    {
        $this->package('spatie/analytics-reports');
    }
    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        $this->app->bindShared('analytics-reports', function ($app) {
            $client = $app->make('analytics');
            $analyticsApi = new AnalyticsReports($client, $app['config']->get('analyticsReports.siteId'), $app['config']->get('analyticsReports.cacheLifetime'));

            return $analyticsApi;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['analytics-reports'];
    }
}


src/AnalyticsReports.php
<?php

namespace Spatie\AnalyticsReports;

use Carbon\Carbon;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Collection;
use Thujohn\Analytics\Analytics;

/**
 * Retrieve data from Google Analytics.
 *
 * Class Analytics
 */
class AnalyticsReports
{
    /**
     * @var Analytics
     */
    protected $client;
    /**
     * @var string
     */
    protected $siteId;
    /**
     * @var int
     */
    protected $cacheLifeTimeInMinutes;

    /**
     * @param Analytics $client                 An already authenticated client
     * @param string    $siteId                 Should look something like ga:xxxxxxxxx
     * @param int       $cacheLifeTimeInMinutes
     */
    public function __construct(Analytics $client, $siteId = '', $cacheLifeTimeInMinutes = 0)
    {
        $this->client = $client;
        $this->siteId = $siteId;
        $this->cacheLifeTimeInMinutes = $cacheLifeTimeInMinutes;
    }

    /**
     * Get the amount of visitors and pageviews.
     *
     * @param int    $numberOfDays
     * @param string $groupBy      Possible values: date, yearMonth
     *
     * @return Collection
     */
    public function getVisitorsAndPageViews($numberOfDays = 365, $groupBy = 'date')
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getVisitorsAndPageViewsForPeriod($startDate, $endDate, $groupBy);
    }

    /**
     * Get the amount of visitors and pageviews for the given period.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param string    $groupBy   Possible values: date, yearMonth
     *
     * @return Collection
     */
    public function getVisitorsAndPageViewsForPeriod($startDate, $endDate, $groupBy = 'date')
    {
        $visitorData = [];
        $answer = $this->performQuery($startDate, $endDate, 'ga:visits,ga:pageviews', ['dimensions' => 'ga:'.$groupBy]);

        foreach ($answer->rows as $dateRow) {
            $visitorData[] = [$groupBy => Carbon::createFromFormat(($groupBy == 'yearMonth' ? 'Ym' : 'Ymd'), $dateRow[0]), 'visitors' => $dateRow[1], 'pageViews' => $dateRow[2]];
        }

        return new Collection($visitorData);
    }

    /**
     * Get the top keywords.
     *
     * @param int $numberOfDays
     * @param int $maxResults
     *
     * @return Collection
     */
    public function getTopKeywords($numberOfDays = 365, $maxResults = 30)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopKeyWordsForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top keywords for the given period.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param int       $maxResults
     *
     * @return Collection
     */
    public function getTopKeyWordsForPeriod($startDate, $endDate, $maxResults = 30)
    {
        $keywordData = [];

        $answer = $this->performQuery($startDate, $endDate, 'ga:sessions', ['dimensions' => 'ga:keyword', 'sort' => '-ga:sessions', 'max-results' => $maxResults, 'filters' => 'ga:keyword!=(not set);ga:keyword!=(not provided)']);

        if (is_null($answer->rows)) {
            return new Collection([]);
        }

        foreach ($answer->rows as $pageRow) {
            $keywordData[] = ['keyword' => $pageRow[0], 'sessions' => $pageRow[1]];
        }

        return new Collection($keywordData);
    }

    /**
     * Get the top referrers.
     *
     * @param int $numberOfDays
     * @param int $maxResults
     *
     * @return Collection
     */
    public function getTopReferrers($numberOfDays = 365, $maxResults = 20)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopReferrersForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top referrers for the given period.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $maxResults
     *
     * @return Collection
     */
    public function getTopReferrersForPeriod($startDate, $endDate, $maxResults)
    {
        $referrerData = [];

        $answer = $this->performQuery($startDate, $endDate, 'ga:pageviews', ['dimensions' => 'ga:fullReferrer', 'sort' => '-ga:pageviews', 'max-results' => $maxResults]);

        if (is_null($answer->rows)) {
            return new Collection([]);
        }

        foreach ($answer->rows as $pageRow) {
            $referrerData[] = ['url' => $pageRow[0], 'pageViews' => $pageRow[1]];
        }

        return new Collection($referrerData);
    }

    /**
     * Get the top browsers.
     *
     * @param int $numberOfDays
     * @param int $maxResults
     *
     * @return Collection
     */
    public function getTopBrowsers($numberOfDays = 365, $maxResults = 6)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopBrowsersForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top browsers for the given period.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $maxResults
     *
     * @return Collection
     */
    public function getTopBrowsersForPeriod($startDate, $endDate, $maxResults)
    {
        $browserData = [];
        $answer = $this->performQuery($startDate, $endDate, 'ga:sessions', ['dimensions' => 'ga:browser', 'sort' => '-ga:sessions']);

        if (is_null($answer->rows)) {
            return new Collection([]);
        }

        foreach ($answer->rows as $browserRow) {
            $browserData[] = ['browser' => $browserRow[0], 'sessions' => $browserRow[1]];
        }

        $browserCollection = new Collection(array_slice($browserData, 0, $maxResults - 1));

        if (count($browserData) > $maxResults) {
            $otherBrowsers = new Collection(array_slice($browserData, $maxResults - 1));
            $otherBrowsersCount = array_sum($otherBrowsers->lists('sessions'));

            $browserCollection->put(null, ['browser' => 'other', 'sessions' => $otherBrowsersCount]);
        }

        return $browserCollection;
    }

    /**
     * Get the most visited pages.
     *
     * @param int $numberOfDays
     * @param int $maxResults
     *
     * @return Collection
     */
    public function getMostVisitedPages($numberOfDays = 365, $maxResults = 20)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getMostVisitedPagesForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the most visited pages for the given period.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param int       $maxResults
     *
     * @return Collection
     */
    public function getMostVisitedPagesForPeriod($startDate, $endDate, $maxResults = 20)
    {
        $pagesData = [];

        $answer = $this->performQuery($startDate, $endDate, 'ga:pageviews', ['dimensions' => 'ga:pagePath', 'sort' => '-ga:pageviews', 'max-results' => $maxResults]);

        if (is_null($answer->rows)) {
            return new Collection([]);
        }

        foreach ($answer->rows as $pageRow) {
            $pagesData[] = ['url' => $pageRow[0], 'pageViews' => $pageRow[1]];
        }

        return new Collection($pagesData);
    }

    /**
     * Returns the site id (ga:xxxxxxx) for the given url.
     *
     * @param $url
     *
     * @throws \Exception
     *
     * @return string
     */
    public function getSiteIdByUrl($url)
    {
        return $this->client->getSiteIdByUrl($url);
    }

    /**
     * Call the query method on the autenthicated client.
     *
     * @param \DateTime $startDate
     * @param \DateTime $endDate
     * @param $metrics
     * @param array     $others
     *
     * @return mixed
     */
    public function performQuery($startDate, $endDate, $metrics, $others = [])
    {
        $startDate = $startDate->format('Y-m-d');
        $endDate = $endDate->format('Y-m-d');
        $cacheName = $this->determineCacheName([$startDate, $endDate, $metrics, $others]);

        if ($this->useCache() and Cache::has($cacheName)) {
            $answer = Cache::get($cacheName);
        } else {
            $answer = $this->client->query($this->siteId, $startDate, $endDate, $metrics, $others);
            if ($this->useCache()) {
                Cache::put($cacheName, $answer, $this->cacheLifeTimeInMinutes);
            }
        }

        return $answer;
    }

    /**
     * Return true if this site is configured to use Google Analytics.
     *
     * @return bool
     */
    public function isEnabled()
    {
        return !$this->siteId == '';
    }

    /**
     * Determine the cache name for the set of query properties given.
     *
     * @param array $properties
     *
     * @return string
     */
    private function determineCacheName(array $properties)
    {
        return 'spatie.analyticsReports.'.md5(serialize($properties));
    }

    /**
     * Returns an array with the current date and the date minus the number of days specified.
     *
     * @param $numberOfDays
     *
     * @return array
     */
    private function calculateNumberOfDays($numberOfDays)
    {
        $endDate = Carbon::today();
        $startDate = Carbon::today()->subDays($numberOfDays);

        return [$startDate, $endDate];
    }

    /**
     * Determine if request to Google should be cached.
     *
     * @return bool
     */
    private function useCache()
    {
        return $this->cacheLifeTimeInMinutes > 0;
    }
}
